src/core/agent/services/keriaNotificationService.ts

```diff
@@ -0,0 +1,743 @@
+        .ipex()
+        .submitAdmit(ourIdentifier.signifyName, admit, sigs, aend, [
+          exchange.exn.i,
+        ]);
+      const pendingOperation = await this.operationPendingStorage.save({
+        id: op.name,
+        recordType: OperationPendingRecordType.ExchangeRevokeCredential,
+      });
+      this.addPendingOperationToQueue(pendingOperation);
+      await this.markNotification(notif.i);
+      return false;
+    }
+    return true;
+  }
+
+  private async processMultiSigRpyNotification(
+    notif: Notification
+  ): Promise<boolean> {
+    const multisigNotification = await this.props.signifyClient
+      .groups()
+      .getRequest(notif.a.d)
+      .catch((error) => {
+        const status = error.message.split(" - ")[1];
+        if (/404/gi.test(status)) {
+          return [];
+        } else {
+          throw error;
+        }
+      });
+    if (!multisigNotification || !multisigNotification.length) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+    const multisigId = multisigNotification[0]?.exn?.a?.gid;
+    if (!multisigId) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+    const multisigIdentifier = await this.identifierStorage
+      .getIdentifierMetadata(multisigId)
+      .catch((error) => {
+        if (
+          error.message === IdentifierStorage.IDENTIFIER_METADATA_RECORD_MISSING
+        ) {
+          return undefined;
+        } else {
+          throw error;
+        }
+      });
+    if (!multisigIdentifier) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+    const rpyRoute = multisigNotification[0].exn.e.rpy.r;
+    if (rpyRoute === "/end/role/add") {
+      await this.multiSigs.joinAuthorization(multisigNotification[0].exn);
+      await this.markNotification(notif.i);
+      return false;
+    }
+    return true;
+  }
+
+  private async processMultiSigIcpNotification(
+    notif: Notification
+  ): Promise<boolean> {
+    const multisigNotification = await this.props.signifyClient
+      .groups()
+      .getRequest(notif.a.d)
+      .catch((error) => {
+        const status = error.message.split(" - ")[1];
+        if (/404/gi.test(status)) {
+          return [];
+        } else {
+          throw error;
+        }
+      });
+    if (!multisigNotification || !multisigNotification.length) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+    const multisigId = multisigNotification[0]?.exn?.a?.gid;
+    if (!multisigId) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+    const hasMultisig = await this.multiSigs.hasMultisig(multisigId);
+    const notificationsForThisMultisig =
+      await this.findNotificationsByMultisigId(multisigId);
+    if (hasMultisig || notificationsForThisMultisig.length) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+    return true;
+  }
+
+  private async processMultiSigExnNotification(
```

Comment:
This function is very different to what's on develop after Bao's recent changes - I think you need to merge develop in or you already have and made some mistakes on the merge conflicts

Reply:
(or git may not have noticed the conflicts)

Reply:
I think it is caused by the renaming

Reply:
Yeah sometimes git thinks its a rename and sometimes thinks it's a delete and new file.

In my experience it behaves differently if you:
- edit the file first and then rename, OR
- rename first and then edit

I can't remember which one is the right way. If I see it's done it incorrectly I usually roll back my changes and do it the other way.

But not too important now, I can do the diff manually. Just grab the latest changes from develop and copy the blocks of code into the file